File: src/java/com/distributedlife/mahjong/reference/hand/HandCandidate.java
package com.distributedlife.mahjong.reference.hand;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class HandCandidate {
    private final String name;
    final List<String> availableTiles;
    private String primarySuit;
    private final List<String> requiredTiles;
    private String secondSuit;
    private String thirdSuit;

    public HandCandidate(String name, List<String> availableTiles) {
        this.name = name;
        this.availableTiles = availableTiles;
        this.requiredTiles = new ArrayList<String>();
        this.secondSuit = "";
        this.thirdSuit = "";
    }

    public HandCandidate(HandCandidate other) {
        this.name = other.name;
        this.primarySuit = other.primarySuit;
        this.availableTiles = new ArrayList<String>(other.availableTiles);
        this.requiredTiles = new ArrayList<String>(other.requiredTiles);
        this.secondSuit = other.secondSuit;
        this.thirdSuit = other.thirdSuit;
    }

    public void setPrimarySuit(String primarySuit) {
        this.primarySuit = primarySuit;
    }

    public boolean isValid() {
        return requiredTiles.size() == 14;
    }

    public String getName() {
        return name;
    }

    public List<String> getRequiredTiles() {
        return requiredTiles;
    }

    public String getPrimarySuit() {
        return primarySuit;
    }

    public boolean canAdd(String tile) {
        return canAdd(tile, times(1));
    }

    public boolean canAdd(String tile, int times) {
        List<String> copyOfRemainingTiles = new ArrayList<String>(availableTiles);

        for (int i = 0; i < times; i++) {
            if (!copyOfRemainingTiles.contains(tile)) {
                return false;
            }

            copyOfRemainingTiles.remove(tile);
        }

        return true;
    }

    public void add(String tile) {
        add(tile, times(1));
    }

    public void add(String tile, int times) {
        for (int i = 0; i < times; i++) {
            availableTiles.remove(tile);
            requiredTiles.add(tile);
        }

        Collections.sort(requiredTiles);
    }

    public boolean equals(Object obj) {
        if (obj == null)
            return false;
        if (obj == this)
            return true;
        if (!(obj instanceof HandCandidate))
            return false;

        HandCandidate rhs = (HandCandidate) obj;
        return new EqualsBuilder().
                append(name, rhs.name).
                append(requiredTiles, rhs.requiredTiles).
                isEquals();
    }

    public int hashCode() {
        return new HashCodeBuilder(17, 31).
                append(name).
                append(requiredTiles).
                toHashCode();
    }

    public static int times(int i) {
        return i;
    }

    public HandCandidate fork() {
        return new HandCandidate(this);
    }

    public String getSecondSuit() {
        return secondSuit;
    }

    public void setSecondSuit(String secondSuit) {
        this.secondSuit = secondSuit;
    }

    public String getThirdSuit() {
        return thirdSuit;
    }

    public void setThirdSuit(String thirdSuit) {
        this.thirdSuit = thirdSuit;
    }
}

File: lib/hand/hand-candidate.js
/**
 * HandCandidate - Intermediate representation during hand generation
 */

export const times = (i) => i;

export class HandCandidate {
  constructor(name, availableTiles, other = null) {
    if (other) {
      // Copy constructor
      this.name = other.name;
      this.primarySuit = other.primarySuit;
      this.availableTiles = [...other.availableTiles];
      this.requiredTiles = [...other.requiredTiles];
      this.secondSuit = other.secondSuit;
      this.thirdSuit = other.thirdSuit;
    } else {
      this.name = name;
      this.availableTiles = availableTiles;
      this.requiredTiles = [];
      this.secondSuit = '';
      this.thirdSuit = '';
    }
  }

  setPrimarySuit(primarySuit) {
    this.primarySuit = primarySuit;
  }

  isValid() {
    return this.requiredTiles.length === 14;
  }

  getName() {
    return this.name;
  }

  getRequiredTiles() {
    return this.requiredTiles;
  }

  getPrimarySuit() {
    return this.primarySuit;
  }

  getSecondSuit() {
    return this.secondSuit;
  }

  setSecondSuit(secondSuit) {
    this.secondSuit = secondSuit;
  }

  getThirdSuit() {
    return this.thirdSuit;
  }

  setThirdSuit(thirdSuit) {
    this.thirdSuit = thirdSuit;
  }

  canAdd(tile, count = times(1)) {
    const copyOfRemainingTiles = [...this.availableTiles];

    for (let i = 0; i < count; i++) {
      const index = copyOfRemainingTiles.indexOf(tile);
      if (index === -1) {
        return false;
      }
      copyOfRemainingTiles.splice(index, 1);
    }

    return true;
  }

  add(tile, count = times(1)) {
    for (let i = 0; i < count; i++) {
      const index = this.availableTiles.indexOf(tile);
      if (index !== -1) {
        this.availableTiles.splice(index, 1);
      }
      this.requiredTiles.push(tile);
    }

    this.requiredTiles.sort();
  }

  equals(obj) {
    if (!obj || !(obj instanceof HandCandidate)) {
      return false;
    }

    return this.name === obj.name &&
           this.primarySuit === obj.primarySuit &&
           this.secondSuit === obj.secondSuit &&
           this.thirdSuit === obj.thirdSuit &&
           JSON.stringify(this.requiredTiles) === JSON.stringify(obj.requiredTiles);
  }

  hashCode() {
    let hash = 17;
    hash = hash * 31 + this.name.split('').reduce((acc, c) => acc + c.charCodeAt(0), 0);
    hash = hash * 31 + this.primarySuit.split('').reduce((acc, c) => acc + c.charCodeAt(0), 0);
    hash = hash * 31 + this.secondSuit.split('').reduce((acc, c) => acc + c.charCodeAt(0), 0);
    hash = hash * 31 + this.thirdSuit.split('').reduce((acc, c) => acc + c.charCodeAt(0), 0);
    hash = hash * 31 + JSON.stringify(this.requiredTiles).split('').reduce((acc, c) => acc + c.charCodeAt(0), 0);
    return hash;
  }

  fork() {
    return new HandCandidate(null, null, this);
  }
}
